patches/server/0739-Add-PlayerCompostItemEvent.patch

```diff
@@ -0,0 +1,59 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Noah van der Aa <ndvdaa@gmail.com>
+Date: Sat, 7 Aug 2021 15:11:33 +0200
+Subject: [PATCH] Add PlayerCompostItemEvent.
+
+
+diff --git a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+index 4c9ae6bdb2f0358798f84928271a2d783dcba7b4..109f17ec2928f01cef405b2235ef16d8946344f9 100644
+--- a/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
++++ b/src/main/java/net/minecraft/world/level/block/ComposterBlock.java
+@@ -219,7 +219,19 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+ 
+         if (i < 8 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
+             if (i < 7 && !world.isClientSide) {
+-                BlockState iblockdata1 = ComposterBlock.addItem(state, (LevelAccessor) world, pos, itemstack);
++                // Paper start
++                float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
++                boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && world.getRandom().nextDouble() >= (double) f); // Paper - Logic moved  up.
++
++                org.bukkit.Location composterLocation = net.minecraft.server.MCUtil.toLocation(world, pos);
++                com.destroystokyo.paper.event.player.PlayerCompostItemEvent event = new com.destroystokyo.paper.event.player.PlayerCompostItemEvent((org.bukkit.entity.Player) player.getBukkitEntity(), composterLocation.getBlock(), itemstack.getBukkitStack(), willRaiseLevel, i);
```

Comment:
Why is this here instead of in `addItem`? Moving it to `addItem` would negate the need to move the "random" logic up.

Reply:
This was moved up so the event could know whether or not the level would be raised. If I were to fire the event in `addItem`, I would have to move the code for decreasing the item, awarding a stat, etc. down aswell, this felt like a better choice.

Reply:
Take a look at [what I did here](https://gist.github.com/Machine-Maker/8053a26c555d8b300239dc0e6fc08bb5).